compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+      } else {
+        // TODO: don't lump in HOF and chains
+        throw new Error(`unexpected token: expected variable or value, found ${work.t.trim()}`);
+      }
+    }
+
+    return stmts;
+  }
+}
+
+export class Assign extends Stmt {
+  upstream: Dec
+  val: Stmt
+
+  constructor(
+    ast: LPNode,
+    upstream: Dec,
+    assignTo: Stmt,
+  ) {
+    super(ast);
+    this.upstream = upstream;
+    this.val = assignTo;
+  }
+
+  static fromAssignmentsAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts: Stmt[] = [];
+    const name = ast.get('varn').t;
+    const upstream = metadata.var(name);
+    if (upstream === null) {
+      throw new Error(`can't assign to ${name}: not found`);
+    }
+    stmts.push(...Stmt.fromAssignables(ast.get('assignables'), metadata));
+    const expr = stmts.pop();
+    let assign: Assign;
+    if (expr instanceof Dec) {
+      assign = new Assign(ast, upstream, expr.val);
+    } else if (expr instanceof Ref) {
+      TODO('figure out how to resolve this???');
+    } else {
+      throw new Error(`invalid assignment state: not a declaration`);
+    }
+    upstream.ty.constrain(assign.val.exprTy());
+    stmts.push(assign);
+    return stmts;
+  }
+
+  exprTy(): Type {
+    throw new Error(`assignments aren't expressions`);
+  }
+
+  inline(amm: Output) {
+    const name = this.upstream.ammName;
+    const ty = this.upstream.ty.breakdown(); // always use the declaration's type, since it's been reduced.
+    if (this.val instanceof Call) {
+      this.val.inline(amm, name, ty, '');
+    } else if (this.val instanceof Lit) {
+      amm.assign('', name, ty, this.val.val);
+    } else {
+      throw new Error(`Unexpected assignment expression: ${this.val}`);
+    }
+  }
+}
+
+export class Call extends Stmt {
+  fn: Fn
+  args: Ref[]
+
+  constructor(
+    ast: LPNode,
+    fn: Fn,
+    args: Ref[],
+  ) {
+    super(ast)
+    this.fn = fn;
+    this.args = args;
+  }
+
+  static fromAsts(
+    wholeAst: LPNode,
+    fnName: string,
+    fnCallAst: LPNode,
+    metadata: MetaData,
+  ): Stmt[] {
+    let stmts = [];
+
+    fnCallAst = fnCallAst.get('assignablelist');
+    const argAsts: LPNode[] = [
+      fnCallAst.get('assignables'),
+      ...fnCallAst.get('cdr').getAll().map(a => a.get('assignables')),
+    ];
+    const args: Ref[] = argAsts.map(a => {
+      const snap = stmts.length;
+      stmts.push(...Stmt.fromAssignables(a, metadata));
+      if (stmts.length <= snap) {
+        throw new Error(`didn't generate any statements for arg ${a.t.trim()}`);
+      }
+      let dec: Stmt = stmts[stmts.length - 1];
+      if (dec instanceof Dec) {
+        return dec.ref();
+      } else if (dec instanceof Ref) {
+        return stmts.pop() as Ref;
+      } else {
+        console.log(dec);
+        throw new Error(`got unexpected statement for arg ${a.t.trim()}: ${dec}`);
+      }
+    });
+
+    let fns = metadata.scope.deepGet(fnName);
+    if (fns === null) {
+      fns = [];
+    }
```

Comment:
I don't think this is necessary `null instanceof Array` returns false so it shouldn't try to do the length check below.

Reply:
I don't want that to happen though - if there aren't any functions in the module scope there could still be a closure defined in the current fn scope that I'd have to find.

Reply:
Well, I mean the next conditional then checks if the length is zero so you still hit the error path, don't you?

Reply:
yes, but I have the TODO on the line before the empty-array-check that'll possibly populate the 0-length-array with a single value. I just have to complete the `Closure` class before I can implement that `TODO`